Давайте решим следующую задачу.<br>
Необходимо написать робота, который будет скачивать новости с сайта Лента.Ру и фильтровать их в зависимости от интересов пользователя. От пользователя требуется отмечать интересующие его новости, по которым система будет выделять области его интересов.<br>


Начнем с загрузки новостей. Для этого нам потребуется метод requests.get(url). Библиотека requests предоставляет серьезные возможности для загрузки информации из Интернет. Метод get получает URL стараницы и возвращает ее содержимое. В нашем случае результат будет получаться в формате html. <br>
Загрузим необходимые библиотеки.

In [14]:
import requests # Загрузка новостей с сайта.
from bs4 import BeautifulSoup # Превращалка html в текст.
import re # Регулярные выражения.

Теперь попробуем загрузить страницу новостей.

In [18]:
requests.get("http://lenta.ru/")

<Response [200]>

Метод <i>requests.get()</i> возвращает объект Response, который содержит большое количество различной информации о загруженной (или незагруженной) странице. В краткой форме отображается только результат выполения запроса. В нашем случае это 200, нет ошибки.<br> 
Посмотрим что результат содержит еще.

In [40]:
%%time
resp=requests.get("https://nplus1.ru/news/2018/08/28/yandex-self-driving")
#resp=requests.get("https://lenta.ru/news/2018/08/24/clon/")
print("cookies:", resp.cookies)
print("time to download:", resp.elapsed)
print("page encoding", resp.encoding)
print("Server response: ", resp.status_code)
print("Is everything ok? ", resp.ok)
print("Page's URL: ", resp.url)

cookies: <RequestsCookieJar[<Cookie XSRF-TOKEN=eyJpdiI6IkRjNkhFNlZaREtYTEI2bmJGa2hhUEE9PSIsInZhbHVlIjoieXBOam84ZHhuRVpqYWZqTHE3RGtcL1pMdkZ3THU3TDZGbVg5NjNBSFJuVml1MUdjUDYyVHNyVFdiZVwvYUFlSUE4NGFTS3NJa3N3R2hMN0dHQW5pYnBhZz09IiwibWFjIjoiMmM5NTYzOTFkYWFlMTkzN2ExNmQ1OWU5NDgzZDM0MWRlMjU4NWQ0NGRlNDAzYzAyNTE4MzMzMzE2MWE0ZjNlYiJ9 for nplus1.ru/>, <Cookie laravel_session=eyJpdiI6ImYrbEdqQVcyTmcrMWljbG1KY213a1E9PSIsInZhbHVlIjoiSVZoY2dMTVpEdkhOYmthbzFjS3FoUjNDYjJ3ZzRIZER1M2lZbEFuaDFjRFFJaXVGNmthQjF0ZjFjdXNaZm4xMWVUT3VUNjhHaEFqelZFY2hEeDZKakE9PSIsIm1hYyI6IjE0YjgxN2JlZDJkOGE3MDE3YjIwMjcwY2U0YTM4ZTRiOTY5YzIwZjhjYWNiZjcxMzZjZjllZmNmYzZlNmE5ODUifQ%3D%3D for nplus1.ru/>]>
time to download: 0:00:00.231685
page encoding UTF-8
Server response:  200
Is everything ok?  True
Page's URL:  https://nplus1.ru/news/2018/08/28/yandex-self-driving
CPU times: user 28.1 ms, sys: 3.81 ms, total: 31.9 ms
Wall time: 240 ms


Но самое для нас интересное хранится в поле <i>text</i>, которое содержит собственно текст html-страницы.

In [41]:
resp.text[:1000]

'<!doctype html>\n<html class="no-js bg-fixed _no-bg" style="background-image:url(https://nplus1.ru/images/2018/08/28/d1a45e4b99c9a988a07ab60414233ab0.jpeg)" lang="">\n<head>\n    <meta charset="utf-8">\n    <meta http-equiv="x-ua-compatible" content="ie=edge">\n    <meta name="viewport" content="width=device-width, initial-scale=1">\n    <link rel="apple-touch-icon" href="apple-touch-icon.png">\n    <meta name="apple-mobile-web-app-capable" content="yes">\n    <meta name="apple-mobile-web-app-status-bar-style" content="black">\n            <meta name="yandex-verification" content="5cb1f59ae9eddd88" /> \n    \n    <link rel="canonical" href="http://nplus1.ru/news/2018/08/28/yandex-self-driving" />\n\n        <title>&laquo;Яндекс&raquo; запустил в Иннополисе сервис беспилотного такси</title>\n\n    \t    <meta itemprop="datePublished" content="2018-08-28"/>\n\t\n\t    <meta name="mediator_author" content="Григорий Копиев"/> \n\t\n\n        <!-- for Google -->\n        <meta name="descri

Количество служебной информации в странице явно превышает объем текста новости. У нас есть два пути: либо использовать библиотеку BeautyfulSoup для получения текста статьи, либо получить текст с использованием регулярных выражений.<br>
Опробуем первый путь. Документация на библиотеку BeautyfulSoup находится <a href="https://www.crummy.com/software/BeautifulSoup/bs4/doc/">здесь</a>.

In [42]:
BeautifulSoup(resp.text, "html5lib").get_text()

'\n    \n    \n    \n    \n    \n    \n             \n    \n    \n\n        «Яндекс» запустил в Иннополисе сервис беспилотного такси\n\n    \t    \n\t\n\t     \n\t\n\n        \n        \n        \n        \n        \n\n        \n        \n                \n        \n        \n        \n        \n        \n        \n        \n        \n        \n\n    \n    \n    \n    \n    \n    \n    \n\n         \n    \n             \n\t\t \n                \t    \n\t\t\n\t\t\n\t\t\n\t\t    var __colAdLoaded = false;\n\t\t    var __topAdLoaded = false;\n\t\t    \n\t\t    var __id = "";\n\t\t    \n\t\t    function detectAd(id){\n\t\t        var _ad = document.getElementById("AdFox_banner_"+id);\n\t\t    \t    console.debug(_ad);\n\t\t        \n\t\t        __id = id;\n\t\t        if(_ad != null && _ad.innerHTML == ""){\n\t\t            $(".col-ad").remove();\n\t\t        }\n\t\t        __colAdLoaded = true;\n\t\t    }\n\t\t    function detectAdTop(id){\n\t\t        __topAdLoaded = true;\n\t\t    }\n\t

Да, убрать html-теги получилось. Но их содержимое осталось, в том числе и скрипты.<br>
Опробуем другой путь. Весь текст обычно оформляется тегом параграфа - &lt;p&gt;. Выберем весь текст из этих тегов. Заодно выберем и заголовок статьи, оформленный при помощи <h1>

In [43]:
bs=BeautifulSoup(resp.text, "html5lib") 
title=bs.h1.text
text=BeautifulSoup(" ".join([p.text for p in bs.find_all("p")]), "html5lib").get_text()
print(title, "\n-----\n", text)

«Яндекс» запустил в Иннополисе сервис беспилотного такси 
-----
 Транспорт
                                             
            
                
                    11:43
                    28 Авг. 2018
                
            
         
            
                Сложность
                1.8
            
          Яндекс Компания «Яндекс» запустила в городе Иннополис в Татарстане тестовый сервис беспилотного такси. Жители города могут вызвать один из двух автомобилей через бота в Telegram и перемещаться между пятью точками высадки и посадки. Во время поездок на переднем пассажирском сиденье будет присутствовать инженер компании, сообщается в блоге «Яндекса». Большинство крупных разработчиков беспилотных автомобилей, к примеру, Waymo и Uber, изначально собираются использовать их в качестве автомобилей для сервисов такси. Во многом это обусловлено тем, что, как правило, в беспилотных автомобилях установлено сразу несколько дорогих лидаров, а также мощные вычислительные мо

Получилось хорошо. Но опробуем второй путь.<br>
Теперь попробуем использовать регулярные выражения в два шага. На первом мы вырежем только саму новость с ее оформлением используя для этого регулярные выражения (библиотека re). На втором шаге мы используем библиотеку BeautifulSoup для "выкусывания" тегов html.

In [44]:
findheaders = re.compile("<h1.+?>(.+)</h1>", re.S)
boa = re.compile('<div class="b-text clearfix js-topic__text" itemprop="articleBody">', re.S)
eoa = re.compile('<div class="b-box">\s*?<i>', re.S)
delscript = re.compile("<script.*?>.+?</script>", re.S)

# Получает текст страницы.
art=requests.get("https://nplus1.ru/news/2018/08/28/yandex-self-driving")
# Находим заголовок.
title = findheaders.findall(art.text)[0]
# Выделяем текст новости.
text = eoa.split(boa.split(art.text)[1])
# Иногда новость оканчивается другим набором тегов.
if len(text)==1:
    text = re.split('<div itemprop="author" itemscope=""', text[0])
# Выкусываем скрипты - BeautifulSoup не справляетсяя с ними.
text = "".join(delscript.split(text[0]))
# Выкусываем остальные теги.
print(BeautifulSoup(title+"\n-----\n"+text,"html5lib").get_text())


IndexError: list index out of range

Обратите внимание на этот фрагмент.<br>
<i>... видимых повреждений на теле не было.У трупа были взяты образцы шерсти...</i><br>
BeautyfulSoup именно "выкусывает" теги, не заменяя их на пробелы. Иногда это можжет приводить к искожению текста из-за "склеивания" слов.

<font color="red" size="6">Закрепим пройденный материал выполнением небольшого задания.</text>

Теперь напишем функцию, которая выгружает все новости за сутки. <br>
Обратим внимание, что для сайта Lenta.ru можно написать адрес в формате lenta.ru/ГГГГ/ММ/ДД/ (год, месяц, день) и получить все новости за этот день. Попробуем получить все адреса с такой страницы.

In [47]:
BeautifulSoup(requests.get("http://lenta.ru/2018/08/25/").text, "html5lib").find_all("a")

[<a class="b-sidebar-menu__auth-link" href="/auth/signin/"><div class="b-sidebar-menu__auth-avatar"><svg class="svg-icons__user b-sidebar-menu__auth-avatar-icon"><use xlink:href="#ui-user"></use></svg></div></a>,
 <a href="/" style="color: #ffffff">Главное</a>,
 <a href="/rubrics/russia/" style="color: #ffffff">Россия</a>,
 <a href="/rubrics/world/" style="color: #ffffff">Мир</a>,
 <a href="/rubrics/ussr/" style="color: #ffffff">Бывший СССР</a>,
 <a href="/rubrics/economics/" style="color: #ffffff">Экономика</a>,
 <a href="/rubrics/forces/" style="color: #ffffff">Силовые структуры</a>,
 <a href="/rubrics/science/" style="color: #ffffff">Наука и техника</a>,
 <a href="/rubrics/culture/" style="color: #ffffff">Культура</a>,
 <a href="/rubrics/sport/" style="color: #ffffff">Спорт</a>,
 <a href="/rubrics/media/" style="color: #ffffff">Интернет и СМИ</a>,
 <a href="/rubrics/style/" style="color: #ffffff">Ценности </a>,
 <a href="/rubrics/travel/" style="color: #ffffff">Путешествия</a>,
 <a 

Кажется, это опять немного не то, что нам нужно. Мы получили все ссылки, находящиеся на боковом меню, ссылки на события сегодняшнего дня и другие ненужные нам вещи. <br>
Смотрим в содержимое html-страницы и обращаем внимание, что все интересные нам ссылки оформлены как заголовки третьего уровня - &lt;h3&gt;. Извлечем все такие фрагменты, а потом извлечем собственно адреса, помеченные атрибутом href тега &lt;a&gt;.

In [48]:
h3s=BeautifulSoup(requests.get("http://lenta.ru/2018/08/25/").text, "html5lib").find_all("h3")
links=["http://lenta.ru"+l.find_all("a")[0]["href"] for l in h3s]
print(links)

['http://lenta.ru/articles/2018/08/25/shishkan/', 'http://lenta.ru/photo/2018/08/25/chicken/', 'http://lenta.ru/photo/2018/08/25/ukrparad/', 'http://lenta.ru/articles/2018/08/25/dzyuba/', 'http://lenta.ru/articles/2018/08/25/berlin/', 'http://lenta.ru/photo/2018/08/25/school_children/', 'http://lenta.ru/brief/2018/08/25/slepakovvsshnur/', 'http://lenta.ru/news/2018/08/25/potomu/', 'http://lenta.ru/news/2018/08/25/rastvorova/', 'http://lenta.ru/news/2018/08/25/razvedka/', 'http://lenta.ru/news/2018/08/25/poezd/', 'http://lenta.ru/news/2018/08/25/bolton/', 'http://lenta.ru/news/2018/08/25/nur/', 'http://lenta.ru/news/2018/08/25/firetornado/', 'http://lenta.ru/news/2018/08/25/drake/', 'http://lenta.ru/news/2018/08/25/merkel/', 'http://lenta.ru/news/2018/08/25/serena/', 'http://lenta.ru/news/2018/08/25/python/', 'http://lenta.ru/news/2018/08/25/ukrain/', 'http://lenta.ru/news/2018/08/25/vsetaki_pustili/', 'http://lenta.ru/news/2018/08/25/boycott_curtis/', 'http://lenta.ru/news/2018/08/25/t

Если теперь написать функцию, которая будет перебирать все адреса и получать из них тексты новостей, то мы получим все новости за определенные сутки. Но это мы сделаем чть позже, а пока просто оформим код загрузки статьи в виде функции.

In [49]:
# Загрузка статьи по URL.
def getOneLentaArticle(url):
    """ getLentaArticle gets the body of an article from Lenta.ru"""
    # Получает текст страницы.
    resp=requests.get(url)
    # Загружаем текст в объект типа BeautifulSoup.
    bs=BeautifulSoup(resp.text, "html5lib") 
    # Получаем заголовок статьи.
    aTitle=bs.h1.text.replace("\xa0", " ")
    # Получаем текст статьи.
    anArticle=BeautifulSoup(" ".join([p.text for p in bs.find_all("p")]), "html5lib").get_text().replace("\xa0", " ")
    return aTitle, anArticle


Разобравшись с базовыми библиотеками, перейдем теперь к обработке собственно текстов. Самостоятельно это можно сделать прочитав одну из двух книг: <a href='https://miem.hse.ru/clschool/the_book'>поновее</a> и <a href='http://clschool.miem.edu.ru/uploads/swfupload/files/011a69a6f0c3a9c6291d6d375f12aa27e349cb67.pdf'>постарше</a> (в старой хорошо разобраны классификация и кластеризация, в новой - тематическое моделирование и рядом лежит видео лекций).<br>
Для обработки текста проводится два этапа анализа: <b>графематический</b> (выделение предложений и слов) и <b>морфологический</b> (определение начальной формы слова, его части речи и грамматических параметров). Этап синтаксического анализа мы разбирать не будем, так как его информация требуется не всегда.<br>
Задачей графематического анализа является разделение текста на составные части - врезки, абзацы, предложения, слова. В таких задачах как машинный перевод, точность данного этапа может существенно влиять на точность получаемых результатов. Например, точка, используемая для сокращений, может быть воспринята как конец предложения, что полность разорвет его семантику.<br>
Но в некоторых задачах (например нашей) используется подход <b>"мешок слов"</b> - текст воспринимается как неупорядоченное множество слов, для которых можно просто посчитать их частотность в тексте. Данный подход проще реализовать, для него не нужно делать выделение составных частей текста, а необходимо только выделить слова.  Именно этот подход мы и будем использовать.<br>

In [26]:
import pymorphy2 # Морфологический анализатор.
from collections import Counter # Не считать же частоты самим.
import math # Корень квадратный.

Задачей морфологического анализа является определение начальной формы слова, его части речи и грамматических параметров. В некоторых случаях от слова требуется только начальная форма, в других - только начальная форма и часть речи.<br>
Существует два больших подхода к морфологическому анализу: <b>стемминг</b> и <b>поиск по словарю</b>. Для проведения стемминга оставляется справочник всех окончаний для данного языка. Для пришедшего слова проверяется его окончание и по нему делается прогноз начальной формы и части речи.<br>
Например, мы создаем справочник, в котором записываем все окончания прилагательных: <i>-ому, -ему, -ой, -ая, -ий, -ый, ...</i> Теперь все слова, которые имеют такое окончание будут считаться прилагаельными: <i>синий, циклический, красного, больному</i>. Заодно прилагательными будут считаться причастия (<i>делающий, строившему</i>) и местоимения (<i>мой, твой, твоему</i>). Также не понятно что делать со словами, имеющими пустое окончание. Отдельную проблему составляют такие слова, как <i>стекло, больной, вина</i>, которые могут разбираться несколькими вариантами (это явление называется <b>омонимией</b>). Помимо этого, стеммер может просто откусывать окончания, оставляя лишь псевдооснову.<br>
Большинство проблем здесь решается, но точность работы бессловарных стеммеров находится на уровне 80%. Чтобы повысить точность испольуют морфологический анализ со словарем. Разработчики составляют словарь слов, встретившихся в текстах (<a href="http://opencorpora.org/dict.php">здесь</a> можно найти пример такого словаря). Теперь каждое слово будет искаться в словаре и не предсказываться, а выдаваться точно. Для слов, отсутствующих в словаре, может применяться предсказание, пообное работе стеммера.<br>
Посмотрим как работает словарная морфология на примере системы pymorphy2.

In [51]:
morph=pymorphy2.MorphAnalyzer() # Создает объект морфоанализатора и загружет словарь.
wordform=morph.parse('стекло')  # Проведем анализ слова "стекло"...
print(wordform)                 # ... и посмотрим на результат.

[Parse(word='стекло', tag=OpencorporaTag('NOUN,inan,neut sing,nomn'), normal_form='стекло', score=0.75, methods_stack=((<DictionaryAnalyzer>, 'стекло', 545, 0),)), Parse(word='стекло', tag=OpencorporaTag('NOUN,inan,neut sing,accs'), normal_form='стекло', score=0.1875, methods_stack=((<DictionaryAnalyzer>, 'стекло', 545, 3),)), Parse(word='стекло', tag=OpencorporaTag('VERB,perf,intr neut,sing,past,indc'), normal_form='стечь', score=0.0625, methods_stack=((<DictionaryAnalyzer>, 'стекло', 968, 3),))]


Как видно из вывода, слово "стекло" может быть неодушевленным существительным среднего рода, единственного числа, именительного падежа <i>tag=OpencorporaTag('NOUN,inan,neut sing,nomn')</i>, аналогично, но в винительном падеже (<i>'NOUN,inan,neut sing,accs'</i>), и глаголом <i>'VERB,perf,intr neut,sing,past,indc'</i>. При этом в первой форме оно встречается в 75% случаев (<i>score=0.75</i>), во второй в 18,75% случаев (<i>score=0.1875</i>), а как глагол - лишь в 6,25% (<i>score=0.0625</i>). Самым простым видом борьбы с омонимией является выбор нулевого элемента из списка, возвращенного морфологическим анализом. Такой подход дает около 90% точности при выборе начальной формы и до 80% если мы обращаем внимание на грамматические параметры.<br><br>
Вместо Pymorphy можно использовать PyMystem. Его плюсом является тот факт, что он сам проводит графематический анализ и снимает омонимию. Используя функцию lemmatize можно получить набор начальных форм слов. Используя функцию analyze можно получить полную информацию о словах.

In [22]:
import pymystem3
mystem=pymystem3.Mystem()

In [23]:
print(mystem.lemmatize('эти типы стали есть в цеху'))
print(mystem.analyze('эти типы стали есть в цеху'))

['этот', ' ', 'тип', ' ', 'становиться', ' ', 'есть', ' ', 'в', ' ', 'цех', '\n']
[{'analysis': [{'lex': 'этот', 'wt': 1, 'gr': 'APRO=(им,мн|вин,мн,неод)'}], 'text': 'эти'}, {'text': ' '}, {'analysis': [{'lex': 'тип', 'wt': 0.8700298642, 'gr': 'S,муж,неод=(вин,мн|им,мн)'}], 'text': 'типы'}, {'text': ' '}, {'analysis': [{'lex': 'становиться', 'wt': 0.9821285244, 'gr': 'V,нп=прош,мн,изъяв,сов'}], 'text': 'стали'}, {'text': ' '}, {'analysis': [{'lex': 'есть', 'wt': 0.0492236161, 'gr': 'V,несов,пе=инф'}], 'text': 'есть'}, {'text': ' '}, {'analysis': [{'lex': 'в', 'wt': 0.9999917878, 'gr': 'PR='}], 'text': 'в'}, {'text': ' '}, {'analysis': [{'lex': 'цех', 'wt': 1, 'gr': 'S,муж,неод=(дат,ед|местн,ед)'}], 'text': 'цеху'}, {'text': '\n'}]


Но мы будем использовать pymorphy, так как он немного пошустрее.<br><br>

In [24]:
art_title, art_text = getOneLentaArticle("https://lenta.ru/news/2018/02/15/greben/")
print(art_title, "\n-----\n", art_text)

NameError: name 'getOneLentaArticle' is not defined

Для новостной заметки можно составить ее словарь, а также посчитать частоты всех слов. В итоге мы получим представление текста в виде вектора. В этом векторе координаты будут называться по соответствующим словам, а смещение по данной координате будет показывать частота. <br>
При составлении словаря будем учитывать только значимые слова - существительные, прилагательные и глаголы. Помимо этого предусмотрим возможность учитывать часть речи слова, прибавляя ее у начальной форме.<br>
Для разделения текста на слова используем простейший алгоритм: слово - это последовательность букв русского алфавита среди которых может попадаться дефис. 

In [21]:
posConv={'ADJF':'_ADJ','NOUN':'_NOUN','VERB':'_VERB'}
meaningfullPoSes=['ADJF', 'NOUN', 'VERB']

def getArticleDictionary(text, needPos=None):
    words=[a[0] for a in re.findall("([А-ЯЁа-яё]+(-[А-ЯЁа-яё]+)*)", text)]
    reswords=[]

    for w in words:
        wordform=morph.parse(w)[0]
        if wordform.tag.POS in meaningfullPoSes:
            if needPos!=None:
                reswords.append(wordform.normal_form+posConv[wordform.tag.POS])
            else:
                reswords.append(wordform.normal_form)
            
    return Counter(reswords)

stat1=getArticleDictionary(art_text, True)
print(stat1)
    


NameError: name 'art_text' is not defined

Более правильным методом является использование CountVectorizer из sklearn.feature_extraction.text. При помощи функции <i>fit_transform</i> можно получить разреженное представление матрицы частот слов. Основная проблема состоит в том, что индексы в матрице представляют собой индексы в словаре переданных текстов. Сам словарь хранится в свойстве <i>vocabulary_</i> и умеет возвращать индекс по слову (но не наоборот).

In [20]:
from sklearn.feature_extraction.text import CountVectorizer

counter=CountVectorizer()
# Просим посчитать частоты слов.
res=counter.fit_transform([art_text])
# Разреженное представление счетчика.
print(res[0][0,:10])
# Можно получить индекс по слову, ...
print(counter.vocabulary_.get('фильм'))
# ... но не наоборот.
print(counter.vocabulary_.get(110))


NameError: name 'art_text' is not defined

Более того, CountVectorizer просто выделяет подстроки и ничего не знает про морфологию (ее можно правильно прикрутить, но это хлопотное занятие). Зато он умеет выделять n-граммы (n слов идущих подряд (или даже букв)). Помимо этого, можно попросить выдать все подстроки, создав анализатор. И можно сказать как выделять подстроки при помощи регулярного выражения.

In [19]:
lyrics="""У тебя в кармане
Два мелка и волшебный камень.
Ты волшебный камень
На Восьмое марта подаришь маме.
А с высокой крыши
Все на свете слышно.
Кто-то хитрый и большой
Наблюдает за тобой."""

# При помощи ngram_range=(1,2) говорим, что хотим извлекать слова и пары слов.
# token_pattern показывает регулярное выражение, которому должны соответствовать слова.
counter12=CountVectorizer(ngram_range=(1,2))#, token_pattern=r'[А-Яа-яЁё]+\-[А-Яа-яЁё]+|[А-Яа-яЁё]+')
# Проводим анализ, получаем список найденных n-грамм.
analyze = counter12.build_analyzer()
print(analyze(lyrics))
# Считаем частоты, видим, что слова не приводились к начальной форме.
res=counter12.fit_transform([lyrics])
print(counter12.vocabulary_.get('маме'))
print(counter12.vocabulary_.get('мама'))

NameError: name 'CountVectorizer' is not defined

Давайте попробуем прикрутить к нему морфологию, но сделаем это немного по-своему. Разобьем строку на слова, проведем их морфологический анализ, возьмем начальные формы и снова склеим в строку, которую отдадим в CountVectorizer. И сделаем тоже самое, только для значимых слов. Как видно из результата, текст перестал быть про "тебя", а стал о чем-то немного другом.

In [18]:
def getMeaningfullWords(text):
    words=[]
    tokens=re.findall('[А-Яа-яЁё]+\-[А-Яа-яЁё]+|[А-Яа-яЁё]+', text)
    for t in tokens:
        pv=morph.parse(t)
        for p in pv:
            if p.tag.POS in ['ADJF', 'NOUN', 'VERB']:
                words.append(p.normal_form)
                break
    return words

c=[' '.join(getMeaningfullWords(lyrics))]
c2=[' '.join([morph.parse(r)[0].normal_form for r in re.findall('[А-Яа-яЁё]+\-[А-Яа-яЁё]+|[А-Яа-яЁё]+', lyrics)])]

lemmaCounter=CountVectorizer(ngram_range=(1,2), token_pattern=r'[А-Яа-яЁё]+\-[А-Яа-яЁё]+|[А-Яа-яЁё]+')

analyze = lemmaCounter.build_analyzer()
res1=analyze(c[0])
res2=lemmaCounter.fit_transform(c)
print({w:res2[0][0,lemmaCounter.vocabulary_[w]] for w in lemmaCounter.vocabulary_ if res2[0][0,lemmaCounter.vocabulary_[w]]>1})
print("---")
res1=analyze(c2[0])
res2=lemmaCounter.fit_transform(c2)
print({w:res2[0][0,lemmaCounter.vocabulary_[w]] for w in lemmaCounter.vocabulary_ if res2[0][0,lemmaCounter.vocabulary_[w]]>1})


NameError: name 'lyrics' is not defined

Обратите внимание на слово "исполняющий". В исходном тексте его не было.<br>
Оно появилось так как одним из (не самых вероятных) значений слова "и" является сокращение от "исполняющий". Так как для каждого слова я пытался найти его значение как значимой части речи, союз был отброшен, а вот прилагательное осталось.

Но это лирика, а в новостях получается не так чувствительно. Более того, на одно из первых мест выходит устойчивое словосочетание - "домашний арест".

In [64]:
c=[' '.join(getMeaningfullWords(art_text))]
c2=[' '.join([morph.parse(r)[0].normal_form for r in re.findall('[А-Яа-яЁё]+\-[А-Яа-яЁё]+|[А-Яа-яЁё]+', art_text)])]

lemmaCounter=CountVectorizer(ngram_range=(1,2), token_pattern=r'[А-Яа-яЁё]+\-[А-Яа-яЁё]+|[А-Яа-яЁё]+')

analyze = lemmaCounter.build_analyzer()
res1=analyze(c[0])
res2=lemmaCounter.fit_transform(c)
print({w:res2[0][0,lemmaCounter.vocabulary_[w]] for w in lemmaCounter.vocabulary_ if res2[0][0,lemmaCounter.vocabulary_[w]]>1})
print("---")
res1=analyze(c2[0])
res2=lemmaCounter.fit_transform(c2)
print({w:res2[0][0,lemmaCounter.vocabulary_[w]] for w in lemmaCounter.vocabulary_ if res2[0][0,lemmaCounter.vocabulary_[w]]>1})


{'группа': 2, 'гребенщик': 2, 'фильм': 3, 'серебренников': 3, 'лето': 3, 'его': 3, 'век': 6, 'февраль': 2, 'сценарий': 3, 'под': 2, 'домашний': 2, 'арест': 2, 'иза': 2, 'науменко': 2, 'зверь': 2, 'год': 2, 'под домашний': 2, 'домашний арест': 2}
---
{'группа': 2, 'гребенщик': 2, 'не': 3, 'фильм': 3, 'серебренников': 3, 'лето': 2, 'он': 3, 'в': 6, 'февраль': 2, 'сценарий': 3, 'под': 2, 'домашний': 2, 'арест': 2, 'из': 2, 'науменко': 2, 'зверь': 2, 'год': 2, 'под домашний': 2, 'домашний арест': 2}


Но нам необходимо искать новости, которые интересны пользователю.<p>
Для определения меры сходства двух статей теперь может использоваться косинусная мера сходства, рассчитываемая по следующей формуле: $cos(a,b)=\frac{\sum{a_i * b_i}}{\sqrt {\sum{a_i^2}*\sum{b_i^2}}}$.<br>
Вообще-то, использовать стандартную функцию рассчета косинусной меры сходства из <a href="http://scikit-learn.org/stable/modules/generated/sklearn.metrics.pairwise.cosine_similarity.html">sklearn</a> было бы быстрее. Но в данной задаче нам бы пришлось сводить все словари в один, чтобы на одних и тех же местах в векторе были частоты одних и тех же слов. Чтобы избежать подобной работы, напишем собственную функцию рассчета косинусного расстояния, работающую с разреженными векторами в виде питоновских словарей.

In [65]:
def cosineSimilarity(a, b):
    if len(a.keys())==0 or len(b.keys())==0:
        return 0
    sumab=sum([a[na]*b[na] for na in a.keys() if na in b.keys()])
    suma2=sum([a[na]*a[na] for na in a.keys()])
    sumb2=sum([b[nb]*b[nb] for nb in b.keys()])
    return sumab/math.sqrt(suma2*sumb2)


Кстати, вы обратили внимание на красивую формулу в маркдауне выше? Здесь можно использовать html и LaTex для красивого оформления. Например, тег &lt;b&gt;позволяет сделать текст <b>полужирным</b>.

Посчитаем значение косинусной меры для разных статей.

In [66]:
stat2=getArticleDictionary(getOneLentaArticle("https://lenta.ru/news/2018/02/15/pengilly_domoi/")[1], True)
stat3=getArticleDictionary(getOneLentaArticle("https://lenta.ru/news/2018/02/15/tar_mor/")[1], True)
stat4=getArticleDictionary(getOneLentaArticle("https://lenta.ru/news/2018/02/15/olympmovies/")[1], True)

print(cosineSimilarity(stat1, stat2))
print(cosineSimilarity(stat1, stat3))
print(cosineSimilarity(stat2, stat3))
print(cosineSimilarity(stat2, stat4))
print(cosineSimilarity(stat3, stat4))

0.0800246267513914
0.0
0.09061004703659373
0.09777979218062152
0.08519427513705972


In [79]:
# training - my function 5/15 пересечение словрей делим на объединение словарей
def closeSimilarity(a, b):
    if len(a.keys())==0 or len(b.keys())==0:
        return 0
    sumab=sum([a[na]*b[na] for na in a.keys() if na in b.keys()])
    suma2=sum([a[na]*a[na] for na in a.keys()])
    sumb2=sum([b[nb]*b[nb] for nb in b.keys()])
#    return sumab/(suma2+sumb2)
    return len(set(a.keys()&set(b.keys())))/len(set(a.keys()|set(b.keys())))
# return len(set(a.keys()&set(b.keys())))/len(set(a.keys))
# return len(set(a.keys()&set(b.keys())))/len(set(a.keys()|set(b.keys())))

In [80]:
print(closeSimilarity(stat1, stat2))
print(closeSimilarity(stat1, stat3))
print(closeSimilarity(stat2, stat3))
print(closeSimilarity(stat2, stat4))
print(closeSimilarity(stat3, stat4))

0.0392156862745098
0.0
0.056
0.04216867469879518
0.050724637681159424


Получилось, на самом деле, так себе - статьи очень слабо походят друг на друга. Но может быть потом выйдет лучше.<br>

<font color="red" size="6">А теперь немного поупражняемся.</font><br>
Напишите функцию, которая ищет процент пересечения словарей у двух статей с применением MyStem.

Оформим теперь весь необходимый код как класс, который будет загружать новости с сайта, сохранять их в файл и читать из предварительно сохраненного файла.<br>
Класс - это тип, определенный пользователем (программистом). Класс содержит в себе как данные, так и функции, которые работают с этими данными.<br>
Так как это тип, то можно создавать переменные этого типа. Каждая переменная будет хранить и обрабатывать свой набор данных.

In [16]:
class getNewsPaper:
        
    # Конструктор - вызывается при создании объекта и инициализирует его.
    def __init__(self):
        self.articles=[]     # Загруженные статьи.
        self.titles=[]       # Заголовки статей.
        self.dictionaries=[] # Словари для каждой из статей.
        # Создаем и загружаем морфологический словарь.
        self.morph=pymorphy2.MorphAnalyzer()

    # Загрузка статьи по URL.
    def getLentaArticle(self, url):
        """ getLentaArticle gets the body of an article from Lenta.ru"""
        # Получает текст страницы.
        resp=requests.get(url)
        # Загружаем текст в объект типа BeautifulSoup.
        bs=BeautifulSoup(resp.text, "html5lib") 
        # Получаем заголовок статьи.
        self.titles.append(bs.h1.text.replace("\xa0", " "))
        # Получаем текст статьи.
        self.articles.append(BeautifulSoup(" ".join([p.text for p in bs.find_all("p")]), "html5lib").get_text().replace("\xa0", " "))

    # Загрузка всех статей за один день.
    def getLentaDay(self, url):
        """ Gets all URLs for a given day and gets all texts. """
        try:
            # Грузим страницу со списком всех статей.
            day = requests.get(url) 
            # Получаем фрагменты с нужными нам адресами статей.
            h3s=BeautifulSoup(day.text, "html5lib").find_all("h3")
            # Получаем все адреса на статьи за день.
            links=["http://lenta.ru"+l.find_all("a")[0]["href"] for l in h3s]
            # Загружаем статьи.
            for l in links:
                self.getLentaArticle(l)
        except:
            pass

    # Загрузка всех статей за несколько дней.
    def getLentaPeriod(self, start, finish):
        curdate=start
        while curdate<=finish:
            print(curdate.strftime('%Y/%m/%d')) # Just in case.
            # Список статей грузится с вот такого адреса.
            self.getLentaDay('https://lenta.ru/news/'+curdate.strftime('%Y/%m/%d'))
            curdate+=datetime.timedelta(days=1)

    # Потроение вектора для статьи.
    posConv={'ADJF':'_ADJ','NOUN':'_NOUN','VERB':'_VERB'}
    def getArticleDictionary(self, text, needPos=None):
        words=[a[0] for a in re.findall("([А-ЯЁа-яё]+(-[А-ЯЁа-яё]+)*)", text)]
        reswords=[]
    
        for w in words:
            wordform=self.morph.parse(w)[0]
            try:
                if wordform.tag.POS in ['ADJF', 'NOUN', 'VERB']:
                    if needPos!=None:
                        reswords.append(wordform.normal_form+self.posConv[wordform.tag.POS])
                    else:
                        reswords.append(wordform.normal_form)
            except:
                pass
            
        stat=Counter(reswords)
#        stat={a: stat[a] for a in stat.keys() if stat[a]>1}
        return stat

    # Посчитаем вектора для всех статей.
    def calcArticleDictionaries(self, needPos=None):
        self.dictionaries=[]
        for a in self.articles:
            self.dictionaries.append(self.getArticleDictionary(a, needPos))
            
    # Сохраняем статьи в файл.
    def saveArticles(self, filename):
        """ Saves all articles to a file with a filename. """
        newsfile=open(filename, "w")
        for art in self.articles:
            newsfile.write('\n=====\n'+art)
        newsfile.close()

    # Читаем статьи из файла.
    def loadArticles(self, filename):
        """ Loads and replaces all articles from a file with a filename. """
        newsfile=open(filename, encoding="utf-8")
        text=newsfile.read()
        self.articles=text.split('\n=====\n')[1:]
        for a in self.articles:
            self.titles.append(a.split('\n-----\n')[0])
        newsfile.close()

    # Для удобства - поиск статьи по ее заголовку.
    def findNewsByTitle(self, title):
        if title in self.titles:
            return self.titles.index(title)
        else:
            return -1

def cosineSimilarity(a, b):
    if len(a.keys())==0 or len(b.keys())==0:
        return 0
    sumab=sum([a[na]*b[na] for na in a.keys() if na in b.keys()])
    suma2=sum([a[na]*a[na] for na in a.keys()])
    sumb2=sum([b[nb]*b[nb] for nb in b.keys()])
    return sumab/math.sqrt(suma2*sumb2)

Загрузим статьи.<br>
<b>!!! Настоятельно рекомендую использовать ячейку с загрузкой статей из файла !!!</b>

In [ ]:

# Загрузка статей за заданный период.
# !!! Это рабоатет довольно долго, пользуйтесь сохраненными данными!!!
lenta=getNewsPaper()
lenta.getLentaPeriod(datetime.date(2018, 2, 1), datetime.date(2018, 2, 14))
lenta.saveArticles("lenta2018.txt")
#lenta.loadArticles("lenta2018.txt")
lenta.calcArticleDictionaries()

In [1]:
import datetime

In [6]:
lenta=getNewsPaper()
lenta.loadArticles("lenta2018-2.txt")
lenta.calcArticleDictionaries()

Из чистого любопытства попробуем найти статью, наиболее похожую на данную.

In [7]:
# Конечно же, правильнее делать это через np.argmax().
i1 = 0
maxCos, maxpos = -1, -1
for i in range(len(lenta.articles)):
    if i != i1:
        c = cosineSimilarity(lenta.dictionaries[i1], lenta.dictionaries[i])
        if c>maxCos:
            maxCos, maxpos = c, i
print(lenta.articles[i1].split('\n-----\n')[0])
print(lenta.articles[maxpos].split('\n-----\n')[0])
print(maxCos, maxpos)

Раскрыто происхождение новейшей украинской крылатой ракеты
Россия поставила Украине оружие «сдерживания агрессора»
0.6522716186492549 516


Сходство между статьями достаточно велико. Есть большие шансы за то, что они об одном и том же.<br><br>
Теперь попробуем решить основную задачу.<br>
Пользователь выбирает несколько статей на интересующую его тематику. Пусть это будут олимпиада и выборы.

In [8]:
likesport=['Власти США обвинили МОК и ФИФА в коррупции', 'Пробирки WADA для допинг-проб оказались бракованными', 'Пожизненно отстраненных российских спортсменов оправдали', 'В Кремле порадовались за оправданных российских спортсменов', 'Россия вернется на первое место Олимпиады-2014', 'МОК разочаровало оправдание российских олимпийцев', 'Мутко загрустил после оправдания российских спортсменов', 'Оправданный призер Сочи-2014 призвал «добить ситуацию» с МОК', 'Путин предостерег от эйфории после оправдания российских олимпийцев', 'Родченков не смог вразумительно ответить на вопросы суда', 'Оправданный россиянин позлорадствовал над делившими медали Игр-2014 иностранцами', 'В CAS отказались считать оправданных россиян невиновными', 'Адвокат Родченкова заговорил о смерти чистого спорта после оправдания россиян', 'Американская скелетонистка сочла россиян ушедшими от законного наказания']
likeelect=['Социологи подсчитали планирующих проголосовать на выборах-2018', 'Собчак пообещала дать Трампу пару советов', 'На выборы президента России пойдут почти 80 процентов избирателей', 'Песков вспомнил предупреждение и отказался комментировать поездку Собчак в США', 'Собчак съездила на завтрак с Трампом и разочаровалась', 'Грудинин уступил в популярности КПРФ', 'Собчак потребовала признать незаконной регистрацию Путина на выборах', 'У Грудинина обнаружили два не до конца закрытых счета в Швейцарии и Австрии', 'Грудинин раскрыл историю происхождения дома в Испании', 'Путина зарегистрировали кандидатом в президенты', 'В Кремле отреагировали на слухи о голосовании Путина в Севастополе', 'Коммунистов вновь обвинили в незаконной агитации за Грудинина', 'ЦИК выявила обман со стороны Грудинина', 'Грудинин ответил на претензии ЦИК', 'Жириновский захотел сбросить ядерную бомбу на резиденцию Порошенко']

Теперь объединим все выбранные тексты в один и посчитаем ветор для него. Сделаем это два раза для выбранных тематик.

In [9]:
sporttext=' '.join([lenta.articles[lenta.findNewsByTitle(art)] for art in likesport])
sportdict=lenta.getArticleDictionary(sporttext)
electtext=' '.join([lenta.articles[lenta.findNewsByTitle(art)] for art in likeelect])
electdict=lenta.getArticleDictionary(electtext)
print(sportdict)
#print(electdict)

Counter({'спортсмен': 47, 'игра': 35, 'олимпиада': 34, 'российский': 30, 'быть': 29, 'суд': 29, 'мокнуть': 28, 'россия': 25, 'пожизненный': 24, 'антидопинговый': 24, 'олимпийский': 22, 'сочи': 22, 'решение': 21, 'нарушение': 16, 'правило': 16, 'февраль': 16, 'результат': 16, 'россиянин': 15, 'материал': 14, 'тема': 14, 'апелляция': 14, 'международный': 13, 'комитет': 13, 'спортивный': 13, 'арбитражный': 13, 'отстранение': 13, 'участие': 13, 'победа': 11, 'год': 10, 'который': 10, 'вернуть': 10, 'медаль': 10, 'удовлетворить': 9, 'родченков': 9, 'вопрос': 8, 'заявить': 8, 'слово': 8, 'весь': 8, 'февраляотстоять': 8, 'сочисо': 8, 'снять': 8, 'четверг': 7, 'аннулировать': 7, 'атлет': 7, 'зачёт': 7, 'тот': 7, 'оправдание': 7, 'агентство': 6, 'этот': 6, 'александр': 6, 'виновный': 6, 'награда': 6, 'первое': 6, 'место': 6, 'медальный': 6, 'ряд': 6, 'приводить': 6, 'спорт': 6, 'дисквалификация': 5, 'глава': 5, 'всемирный': 5, 'лыжник': 5, 'дать': 5, 'наш': 5, 'чистый': 5, 'сша': 4, 'расследова

А теперь отберем все статьи, косинусная мера которых превышает некоторый порог.

In [10]:
thrs=0.4
thre=0.5
cosess=[lenta.articles[i].split('\n-----\n')[0] for i in range(len(lenta.dictionaries)) if cosineSimilarity(sportdict, lenta.dictionaries[i])>thrs]
print(cosess)
cosese=[lenta.articles[i].split('\n-----\n')[0] for i in range(len(lenta.dictionaries)) if cosineSimilarity(electdict, lenta.dictionaries[i])>thre]
print(cosese)

['Пожизненно отстраненных российских спортсменов оправдали', 'В Кремле порадовались за оправданных российских спортсменов', 'Россия вернется на первое место Олимпиады-2014', 'МОК разочаровало оправдание российских олимпийцев', 'Мутко загрустил после оправдания российских спортсменов', 'Олимпиада в Пхенчхане побила рекорд по презервативам', 'Оправданный призер Сочи-2014 призвал «добить ситуацию» с МОК', 'Путин предостерег от эйфории после оправдания российских олимпийцев', 'Родченков не смог вразумительно ответить на вопросы суда', 'Оправданный россиянин позлорадствовал над делившими медали Игр-2014 иностранцами', 'В CAS отказались считать оправданных россиян невиновными', 'Адвокат Родченкова заговорил о смерти чистого спорта после оправдания россиян', 'Американская скелетонистка сочла россиян ушедшими от законного наказания', 'Глава USADA почуял вонь российской атаки на чистый спорт', 'После оправдания российских спортсменов Макларена назвали идиотом', 'МОК посчитал оправдание российск

Для проверки загрузим новости за какой-то другой день.

In [11]:
lenta_new=getNewsPaper()
#lenta_new.getLentaPeriod(datetime.date(2018, 2, 15), datetime.date(2018, 2, 15))
#lenta_new.saveArticles("lenta20180215.txt")
lenta_new.loadArticles("lenta20180215.txt")
lenta_new.calcArticleDictionaries()

FileNotFoundError: [Errno 2] No such file or directory: 'lenta20180215.txt'

А теперь проверим какие новости будут находиться.

In [152]:
thrs_new = 0.3
thre_new = 0.3
cosess_new = [lenta_new.articles[i].split('\n-----\n')[0] for i in range(len(lenta_new.dictionaries)) if cosineSimilarity(sportdict, lenta_new.dictionaries[i])>thrs_new]
print(cosess_new)
cosese_new = [lenta_new.articles[i].split('\n-----\n')[0] for i in range(len(lenta_new.dictionaries)) if cosineSimilarity(electdict, lenta_new.dictionaries[i])>thre_new]
print(cosese_new)

['Российский сноубордист сломал ногу на Олимпиаде', 'Российскую лыжницу затравили на Олимпиаде']
['Кремль опроверг информацию о засекречивании данных по россиянам в Сирии', 'Порошенко раскрыл детали разговора с Путиным', 'Грудинин оговорился о финансовых махинациях', 'Прогнозируемую явку на выборах президента сочли высокой']


Как видно, метод нуждается в более точном подборе и корректировке параметров.

Вместо того, чтобы выделять самые частотные слова в одной статье, мы можем использовать несколько иную логику. Слово является важным, если оно часто встречается в <b>данной</b> статье. Однако если это слово часто встречается во всех статьях, скорее всего оно не значимо (так ведут себя предлоги, союзы, слова, относящиеся к особенностям авторского или принятого в данной области стиля). То есть наиболее характерными для данного документа будут слова и словосочетания, которые встречаются часто в данном докумемнте, но в малом количестве документов.<br>
Для выделения таких слов будем использовать TfidfVectorizer, работающий по формуле tf/idf, где tf - term frequency, а idf - inverted document frequency. То есть в простейшем случае можно просто разделить частоту термина в документе на количество документов, в которых он встречается.

In [12]:
from sklearn.feature_extraction.text import TfidfVectorizer
import numpy as np

In [28]:
c=[' '.join(getMeaningfullWords(n)) for n in lenta.articles]

tfCounter=TfidfVectorizer(ngram_range=(1,2), token_pattern=r'[А-Яа-яЁё]+\-[А-Яа-яЁё]+|[А-Яа-яЁё]+')

analyze = tfCounter.build_analyzer()
res=tfCounter.fit_transform(c)

NameError: name 'morph' is not defined

In [155]:
id_article=6

res2=analyze(lenta.articles[id_article])
tfs=list(set(res[id_article][0, tfCounter.vocabulary_.get(k)] for k in res2 if k in tfCounter.vocabulary_.keys()))
tfs2=[k for k in tfs if k>np.average(tfs)]
#tfs2=[k for k in tfs if k>np.average(tfs)+np.std(tfs)]
print({w:res[id_article][0, tfCounter.vocabulary_[w]] for w in res2 if w in tfCounter.vocabulary_.keys() and res[id_article][0, tfCounter.vocabulary_[w]] in tfs2})


{'ближний': 0.09754172040483444, 'восток': 0.10389546512876865, 'платформа': 0.11428148479237515, 'жозе': 0.16535511285226273, 'увеличение': 0.0842033272150634, 'возможность': 0.08574845736217529, 'доставка': 0.13234427079429065, 'клиентский': 0.1744831264097395, 'сервис': 0.0709906385519639, 'услуга': 0.06742359087070612, 'ближний восток': 0.13825073355807999, 'клиентский сервис': 0.11632208427315967}
